### ISSUE TITLE
Internal error with 1.6rc1 caused by incorrect code

### ISSUE THREAD

**Issue by paulmelis** [JuliaLang/julia#39862]
Reduced example for a case where I inadvertently used the wrong call syntax for the `randvect` call, resulting in an internal error. This is with 1.6.0RC1 (see below for versioninfo output)

```
melis@juggle 21:21:~/c/PathTracer.jl$ cat b2.jl 
randvect() = 4

function gen()
    return randvect{Float32}()
end

gen()
```

```
melis@juggle 21:21:~/c/PathTracer.jl$ ~/software/julia-1.6.0-rc1/bin/julia b2.jl 
Internal error: encountered unexpected error in runtime:
MethodError(f=Core.Compiler.:(==), args=(Main.randvect, Tuple), world=0x00000000000010a8)
jl_method_error_bare at /buildworker/worker/package_linux64/build/src/gf.c:1813
jl_method_error at /buildworker/worker/package_linux64/build/src/gf.c:1831
jl_lookup_generic_ at /buildworker/worker/package_linux64/build/src/gf.c:2401 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2416
apply_type_tfunc at ./compiler/tfuncs.jl:1147
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:672
builtin_tfunction at ./compiler/tfuncs.jl:1488
abstract_call_builtin at ./compiler/abstractinterpretation.jl:827
jfptr_abstract_call_builtin_8526.clone_1 at /home/melis/software/julia-1.6.0-rc1/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
abstract_call_known at ./compiler/abstractinterpretation.jl:958
abstract_call at ./compiler/abstractinterpretation.jl:1060
abstract_call at ./compiler/abstractinterpretation.jl:1044
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1171
typeinf_local at ./compiler/abstractinterpretation.jl:1449
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1507
_typeinf at ./compiler/typeinfer.jl:214
typeinf at ./compiler/typeinfer.jl:209
typeinf_ext at ./compiler/typeinfer.jl:892
typeinf_ext_toplevel at ./compiler/typeinfer.jl:925
typeinf_ext_toplevel at ./compiler/typeinfer.jl:921
jfptr_typeinf_ext_toplevel_14033.clone_1 at /home/melis/software/julia-1.6.0-rc1/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
jl_type_infer at /buildworker/worker/package_linux64/build/src/gf.c:298
jl_generate_fptr at /buildworker/worker/package_linux64/build/src/jitlayers.cpp:340
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:1971
jl_compile_method_internal at /buildworker/worker/package_linux64/build/src/gf.c:2237 [inlined]
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2230 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:155 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:561
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:669
top-level scope at /home/melis/c/PathTracer.jl/b2.jl:7
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:879
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:827
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:931
eval at ./boot.jl:360 [inlined]
include_string at ./loading.jl:1090
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
_include at ./loading.jl:1144
include at ./Base.jl:386
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
exec_options at ./client.jl:285
_start at ./client.jl:485
jfptr__start_50107.clone_1 at /home/melis/software/julia-1.6.0-rc1/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
true_main at /buildworker/worker/package_linux64/build/src/jlapi.c:557
repl_entrypoint at /buildworker/worker/package_linux64/build/src/jlapi.c:699
main at /buildworker/worker/package_linux64/build/cli/loader_exe.c:51
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
_start at /home/melis/software/julia-1.6.0-rc1/bin/julia (unknown line)
ERROR: LoadError: TypeError: in Type{...} expression, expected UnionAll, got a value of type typeof(randvect)
Stacktrace:
 [1] gen()
   @ Main ~/c/PathTracer.jl/b2.jl:4
 [2] top-level scope
   @ ~/c/PathTracer.jl/b2.jl:7
in expression starting at /home/melis/c/PathTracer.jl/b2.jl:7
```

```
julia> versioninfo()
Julia Version 1.6.0-rc1
Commit a58bdd9010 (2021-02-06 15:49 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-4460  CPU @ 3.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, haswell)
```

With 1.5.3 the internal error doesn't happen:
```
melis@juggle 21:21:~/c/PathTracer.jl$ j b2.jl 
ERROR: LoadError: TypeError: in Type{...} expression, expected UnionAll, got a value of type typeof(randvect)
Stacktrace:
 [1] gen() at /home/melis/c/PathTracer.jl/b2.jl:4
 [2] top-level scope at /home/melis/c/PathTracer.jl/b2.jl:7
in expression starting at /home/melis/c/PathTracer.jl/b2.jl:7

melis@juggle 21:21:~/c/PathTracer.jl$ j
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  
|__/                   |

julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1* (2020-11-09 13:37 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-4460  CPU @ 3.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-10.0.1 (ORCJIT, haswell)
```

**Comment by JeffBezanson** [comment]
Fixed on master but could be backported to 1.6.

**Comment by ararslan** [comment]
Didn't make it into RC2 but we should figure out what fixed this so that it can make it into 1.6 final.

**Comment by vtjnash** [comment]
The fix was an incidental part of #38136. The PR cannot be backported, but the relevant change can:
```
-    istuple = (headtype == Tuple)
+    istuple = isa(headtype, Type) && (headtype == Tuple)
```
